File: rohit_common/rohit_common/report/file_attachment_analysis_rigpl/file_attachment_analysis_rigpl.js
// Copyright (c) 2016, Rohit Industries Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["File Attachment Analysis RIGPL"] = {
	"filters": [
		{
			"fieldname":"summary_dt",
			"label": "Summary Doctype Wise",
			"fieldtype": "Check",
			"reqd": 0,
			"default":1
		},
		{
			"fieldname":"summary_fol",
			"label": "Summary Folder Wise",
			"fieldtype": "Check",
			"reqd": 0,
			"default":0
		},
		{
			"fieldname":"is_folder",
			"label": "Is Folder",
			"fieldtype": "Check",
			"reqd": 0,
			"default":0
		},
		{
			"fieldname":"archive",
			"label": "Important for Archive",
			"fieldtype": "Check",
			"reqd": 0,
			"default":0
		},
		{
			"fieldname":"deletion",
			"label": "Marked for Deletion",
			"fieldtype": "Check",
			"reqd": 0,
			"default":0
		},
		{
			"fieldname":"no_parent",
			"label": "No Parent",
			"fieldtype": "Check",
			"reqd": 0,
			"default":0
		},
		{
			"fieldname":"on_server",
			"label": "Available on Server",
			"fieldtype": "Select",
			"reqd": 0,
			"options": "\nYes\nNo"
		},
		{
			"fieldname":"private",
			"label": "Private or Public",
			"fieldtype": "Select",
			"options": "\nAll\nOnly Private\nOnly Public",
			"reqd": 1,
			"default": "All"
		},
		{
			"fieldname":"doctype",
			"label": "Attached to Doctype",
			"fieldtype": "Link",
			"options": "DocType",
			"reqd": 0,
			"get_query": function(){ return {'filters': [['DocType', 'istable','!=', 1]]}}
		},
		{
			"fieldname":"dt_types",
			"label": "Attached to Doctypes",
			"fieldtype": "Select",
			"options": "\nAll\nNone",
		},
		{
			"fieldname":"folder",
			"label": "Folder",
			"fieldtype": "Link",
			"options": "File",
			"get_query": function(){ return {'filters': [['File', 'is_folder','=', 1]]}}
		},
	]
}

File: rohit_common/rohit_common/report/file_attachment_analysis_rigpl/file_attachment_analysis_rigpl.py
# Copyright (c) 2013, Rohit Industries Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from ....utils.rohit_common_utils import get_folder_details


def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data


def get_columns(filters):
    if filters.get("summary_dt") == 1:
        return ["Attached to Doctype::300", "No of Files:Int:100", "Size (MB):Float:100"]
    elif filters.get("summary_fol") == 1:
        return ["Folder Name:Link/File:300", "Parent Folder:Link/File:300", "No of Files:Int:100",
            "Actual Files Size (MB):Float:100", "Size Listed:Float:100", "Left:Int:80", "Right:Int:80"]
    else:
        if filters.get("is_folder") == 1:
            return [
                "ID:Link/File:450", "File Name::200", "Parent Folder:Link/File:350", "Size(MB):Float:100",
                "Left:Int:80", "Right:Int:80", "Files:Int:80","Is Home:Int:40", "Is Attachment::40"
            ]
        else:
            return [
                "ID:Link/File:150", "File Name::200",
                {
                    "label": "Attached To DT",
                    "fieldname": "attached_to_dt",
                    "width": 100
                },
                {
                    "label": "Attached to Name",
                    "fieldname": "attached_to_name",
                    "fieldtype": "Dynamic Link",
                    "options": "attached_to_dt",
                    "width": 100
                },
                "Available:Int:50", "Size(kB):Float:100", "Left:Int:80", "Right:Int:80", "Parent Folder:Link/File:350",
                "Private:Int:40", "Imp:Int:40", "Del:Int:40", "Owner::150", "Created On:Datetime:150", "URL::300"
            ]


def get_data(filters):
    data = []
    conditions, cond_summary = get_conditions(filters)
    if filters.get("summary_dt") == 1:
        data = frappe.db.sql("""SELECT IFNULL(attached_to_doctype, "NO DOCTYPE"), COUNT(name) as no_of_files,
        ROUND(((SUM(file_size))/1024/1024),2) as size FROM `tabFile` WHERE docstatus=0 AND is_folder=0 %s
        GROUP BY attached_to_doctype ORDER BY size DESC, no_of_files DESC """ % cond_summary, as_list=1)
    elif filters.get("summary_fol") == 1:
        new_data = frappe.db.sql("""SELECT name, folder, ROUND(file_size/1024/1024, 2) as size, lft, rgt
            FROM `tabFile`
            WHERE docstatus=0 AND is_folder=1
            ORDER BY lft DESC, rgt DESC""", as_dict=1)
        for row in new_data:
            files = frappe.db.sql("""SELECT ROUND(((SUM(file_size))/1024/1024),2) as size, COUNT(name) as nos
            FROM `tabFile` WHERE folder='%s'""" %row.name, as_dict=1)
            row["act_size"] = files[0].size
            row["count"] = files[0].nos
        for row in new_data:
            data_row = [row.name, row.folder, row.count, row.act_size, row.size, row.lft, row.rgt]
            data.append(data_row)
    else:
        if filters.get("is_folder") == 1:
            data = frappe.db.sql("""SELECT name, file_name, folder, ROUND(file_size/1024/1024,2), lft, rgt, (rgt - lft),
                is_home_folder, is_attachments_folder
                FROM `tabFile` WHERE docstatus = 0 %s ORDER BY lft, rgt""" % (conditions), as_list=1)
        else:
            query = """SELECT name, IFNULL(file_name, "NO NAME") as file_name, IFNULL(attached_to_doctype, "NO DOCTYPE") as atd,
                IFNULL(attached_to_name,"NO DOCNAME") as atn, file_available_on_server,
                ROUND(file_size/1024,2) as size, lft, rgt, IFNULL(folder, "NO FOLDER") as folder, is_private,
                important_document_for_archive, mark_for_deletion, owner, creation, file_url
                FROM `tabFile` WHERE docstatus=0 %s ORDER BY creation""" % (conditions)
            fd_data = frappe.db.sql(query, as_dict=1)
            for d in fd_data:
                file_download_name = """<a href="%s" target="_blank">%s</a>""" % (d.file_url, d.file_name)
                file_download_url = """<a href="%s" target="_blank">%s</a>""" % (d.file_url, d.file_url)
                row = [d.name, file_download_name, d.atd, d.atn, d.file_available_on_server, d.size, d.lft, d.rgt, d.folder,
                    d.is_private, d.important_document_for_archive, d.mark_for_deletion, d.owner, d.creation, file_download_url]
                data.append(row)
    return data


def get_conditions(filters):
    conditions = ""
    cond_summary = ""
    if filters.get("is_folder") == 1:
        conditions += " AND is_folder=1"
    else:
        conditions += " AND is_folder=0"

    if filters.get("private") == "Only Private":
        conditions += " AND is_private=1"
        cond_summary += " AND is_private=1"
    elif filters.get("private") == "Only Public":
        conditions += " AND is_private=0"
        cond_summary += " AND is_private=0"
    else:
        pass

    if filters.get("dt_types") != "None" and filters.get("doctype"):
        conditions += " AND attached_to_doctype = '%s'" % (filters.get("doctype"))
    elif filters.get("dt_types") == "None" and filters.get("doctype"):
        frappe.throw("None Doctype Selected and hence Cannot Select a Specific Doctype")

    if filters.get("dt_types") == "None":
        conditions += " AND attached_to_doctype IS NULL"

    if filters.get("folder"):
        folder_details = get_folder_details(filters.get("folder"))
        lft = folder_details[0].lft + 1
        rgt = folder_details[0].rgt - 1
        conditions += " AND lft >= %s AND rgt <= %s" % (lft, rgt)

    if filters.get("no_parent") == 1:
        conditions += " AND folder IS NULL"

    if filters.get("on_server") == "Yes":
        conditions += " AND file_available_on_server = 1"
        cond_summary += " AND file_available_on_server = 1"
    elif filters.get("on_server") == "No":
        conditions += " AND file_available_on_server = 0"
        cond_summary += " AND file_available_on_server = 0"

    if filters.get("deletion"):
        cond_summary += " AND mark_for_deletion = 1"
        conditions += " AND mark_for_deletion = 1"

    if filters.get("archive"):
        cond_summary += " AND important_document_for_archive = 1"
        conditions += " AND important_document_for_archive = 1"

    return conditions, cond_summary
